### ISSUE TITLE
Investigate why RunSGAViaProcessBuilderOnSpark runs extremely slow with large number of files

### ISSUE THREAD

**Issue by SHuang-Broad** [broadinstitute/gatk#1877]
Reading the same number of files by `JavaSparkContext.wholeTextFiles` and counting the number of lines in each file takes approximately 20 min. But running locally assembly on these files takes more than 24 hours, which based on experience is way too much longer than expected.

In addition, running the same program on ~10 files takes ~5 min. 

Need to see if there's a way to debug this performance issue (get information from executors).

**Comment by cwhelan** [comment]
A couple of things we can try:

Make sure that the temp directories created by RunSGAViaProcessBuilderOnSpark are actually being created on the /tmp filesystem. Maybe there's some Spark configuration that is overriding that?

Try removing the assembly part and just running writeToLocal() on the text files dataset. Then we can see if something's getting stuck in the assembly process or if it's a file-handling problem.

**Comment by SHuang-Broad** [comment]
For @cwhelan and who are interested, testing without modifying logic (but added printf debugging logs) 
seems to suggest that running assembly on 5,000 FASTQ files with sizes less than 280KB takes ~8 hours, on 4 executors (8 cores each), meaning approximately 3~4 minutes for each, which is not too bad.

**Comment by SHuang-Broad** [comment]
Another update: when test run on 10,000 files using 4 executors, one strange observation is that only one executor is doing the work, leading to low efficiency. Not sure exactly why this is the case.
http://dataflow01.broadinstitute.org:18088/history/application_1464285223085_0152/executors/

**Comment by cwhelan** [comment]
Could it be that wholeTextFiles is reading everything into one partition,
and we might need to force a repartition first?

**Comment by SHuang-Broad** [comment]
@cwhelan , you are right! With forced repartition and ~40K files, [this](http://dataflow03.broadinstitute.org:8088/proxy/application_1464285223085_0196/stages/stage?id=2&attempt=0) shows results for one test run.

Just noticed there's another function 

```
public JavaPairRDD<String,String> wholeTextFiles(String path, int minPartitions)
```
will give it a try tomorrow.

**Comment by SHuang-Broad** [comment]
By using 5 executors, 24 cores on each when no one else seem to be using the local cluster, [this](http://dataflow01.broadinstitute.org:18088/history/application_1464285223085_0420/jobs/) is the time it takes to run on the 44188 small FASTQ files, without capturing the stdout and stderr of SGA processes.

Note that this time the RDD caching is done differently than the way it is done in code living in master:

        results.cache(); // cache because Spark doesn't have an efficient RDD.split(predicate) yet
        results.count(); // ugly hack to make the actual computation happen, so later filtering step will be based on what has been actually computed

        // save fasta file contents or failure message
        final JavaPairRDD<Long, SGAAssemblyResult> success = results.filter(entry -> entry._2().assembledContigs!=null);
        final JavaPairRDD<Long, SGAAssemblyResult> failure = results.filter(entry -> entry._2().assembledContigs==null);

        if(!success.isEmpty()){
            success.map(entry -> entry._1().toString() + "\n" + entry._2().assembledContigs.toString())
                    .saveAsTextFile(outputDir+"_0");
        }

        if(!failure.isEmpty()){
            failure.map(entry ->  entry._1().toString() + "\n" + entry._2().collectiveRuntimeInfo.toString())
                    .saveAsTextFile(outputDir+"_1");
        }

**Comment by lbergelson** [comment]
@SHuang-Broad Could you explain what happens with and without the count?  

Spark is deliberately lazy... i.e. without the count you would expect nothing to be computed until success.isEmpty() is called.  (Note that success is a different RDD than results, so calling success.isEmpty() and then success.map(..).saveAsTextFile(..) will probably result in success being re-filtered from results twice.  You could try caching success to avoid that.  In either case it should avoid recomputing results though when failure.isEmpty() is computed.

**Comment by SHuang-Broad** [comment]
@lbergelson , as we discussed, I made the mistake of not using the returned RDD when calling `cache()`, after I fixed it, [here](http://dataflow01.broadinstitute.org:18088/history/application_1464285223085_0460/jobs/) is the runtime, running the following code:

```
148         final JavaPairRDD<Long, SGAAssemblyResult> cachedResults = results.cache(); // cache because Spark doesn't have an efficient RDD.split(predicate) yet
149 //        results.count(); // ugly hack to make the actual computation happen, so later filtering step will be based on what has been actually computed
150 
151         // save fasta file contents or failure message
152         final JavaPairRDD<Long, SGAAssemblyResult> success = cachedResults.filter(entry -> entry._2().assembledContigs!=null);
153         final JavaPairRDD<Long, SGAAssemblyResult> failure = cachedResults.filter(entry -> entry._2().assembledContigs==null);
154 
155         if(!success.isEmpty()){
156             success.map(entry -> entry._1().toString() + "\n" + entry._2().assembledContigs.toString())
157                     .saveAsTextFile(outputDir+"_0");
158         }
159 
160         if(!failure.isEmpty()){
161             failure.map(entry ->  entry._1().toString() + "\n" + entry._2().collectiveRuntimeInfo.toString())
162                     .saveAsTextFile(outputDir+"_1");
163         }
```

Thanks!

**Comment by SHuang-Broad** [comment]
Closed by #1912
